audio/out/ao_avfoundation.m

```diff
@@ -0,0 +1,349 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ao.h"
+#include "audio/format.h"
+#include "audio/out/ao_coreaudio_chmap.h"
+#include "audio/out/ao_coreaudio_utils.h"
+#include "common/common.h"
+#include "common/msg.h"
+#include "internal.h"
+#include "osdep/timer.h"
+#include "ta/ta_talloc.h"
+
+#import <AVFoundation/AVFoundation.h>
+#import <Foundation/Foundation.h>
+#import <CoreAudioTypes/CoreAudioTypes.h>
+#import <CoreFoundation/CoreFoundation.h>
+#import <CoreMedia/CoreMedia.h>
+
+
+@interface AVObserver : NSObject {
+    struct ao *ao;
+}
+- (void)handleRestartNotification:(NSNotification*)notification;
+@end
+
+struct priv {
+    AVSampleBufferAudioRenderer *renderer;
+    AVSampleBufferRenderSynchronizer *synchronizer;
+    dispatch_queue_t queue;
+    CMAudioFormatDescriptionRef format_description;
+    AVObserver *observer;
+    int64_t end_time_av;
+};
+
+static int64_t CMTimeGetNanoseconds(CMTime time)
+{
+    time = CMTimeConvertScale(time, 1000000000, kCMTimeRoundingMethod_Default);
+    return time.value;
+}
+
+static CMTime CMTimeFromNanoseconds(int64_t time)
+{
+    return CMTimeMake(time, 1000000000);
+}
+
+static void feed(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+    CMBlockBufferRef block_buffer = NULL;
+    CMSampleBufferRef sample_buffer = NULL;
+
+    int64_t cur_time_av = CMTimeGetNanoseconds([p->synchronizer currentTime]);
+    int64_t cur_time_mp = mp_time_ns();
+    p->end_time_av = MPMAX(p->end_time_av, cur_time_av);
+
+    int request_sample_count = ao->samplerate / 10; // 1/10 samplerate may be a good number
+    int buffer_size = request_sample_count * ao->sstride;
+    if (CMBlockBufferCreateWithMemoryBlock(
+        kCFAllocatorDefault,
+        NULL,
+        buffer_size,
+        NULL,
+        NULL,
+        0,
+        buffer_size,
+        kCMBlockBufferAssureMemoryNowFlag,
+        &block_buffer
+    ) != noErr) {
+        MP_FATAL(ao, "failed to create CMBlockBuffer\n");
+        goto error;
+    }
+    void* data[1];
+    if (CMBlockBufferGetDataPointer(block_buffer, 0, NULL, NULL, (char**)data) != noErr) {
+        MP_FATAL(ao, "CMBlockBufferGetDataPointer failed\n");
```

Comment:
since those are user facing error, might describe the error instead of returning the actual implementation.

eg "failed to get data from block buffer" or something similar.

Reply:
also if we are printing error messages, please include actual error value, it often gives infinite more insight into the issue when looking at the log from the users.